value.go

```diff
@@ -40,28 +75,131 @@ func (v Value) MatchEqual(other Value) (isMatch bool, isEqual bool) {
 	return
 }
 
-// MarshalValue serializes a single value
-func MarshalValue(value Value) ([]byte, error) {
-	return json.Marshal(&value)
+func (JsonValueSerde) MarshalValue(v Value) ([]byte, error) {
+	return json.Marshal(&v)
 }
 
-func UnmarshalValue(b []byte) (Value, error) {
-	var value Value
-	err := json.Unmarshal(b, &value)
-	return value, err
+func (JsonValueSerde) UnmarshalValue(b []byte) (v Value, err error) {
+	err = json.Unmarshal(b, &v)
+	return
+}
+
+func (JsonValueSerde) MarshalValueKeys(vk [][]byte) ([]byte, error) {
+	return json.Marshal(&vk)
+}
+
+func (JsonValueSerde) UnmarshalValueKeys(b []byte) (vk [][]byte, err error) {
+	err = json.Unmarshal(b, &vk)
+	return
+}
+
+func (BinaryValueSerde) MarshalValue(v Value) ([]byte, error) {
+	var buf []byte
+	pid := []byte(v.ProviderID)
+	buf = append(buf, varint.ToUvarint(uint64(len(pid)))...)
+	buf = append(buf, pid...)
+	buf = append(buf, varint.ToUvarint(uint64(len(v.ContextID)))...)
+	buf = append(buf, v.ContextID...)
+	buf = append(buf, varint.ToUvarint(uint64(len(v.MetadataBytes)))...)
+	buf = append(buf, v.MetadataBytes...)
+	return buf, nil
```

Comment:
Buffer allocations can be avoided by calculating the size needed and then pre-growing the buffer to the necessary size. That should help the benchmarks even more.
```suggestion
	var b bytes.Buffer
	b.Grow(varint.UvarintSize(uint64(len(pid))) + len(pid) +
		varint.UvarintSize(uint64(len(v.ContextID))) + len(v.ContextID) +
		varint.UvarintSize(uint64(len(v.MetadataBytes))) + v.MetadataBytes)
	b.Write(varint.ToUvarint(uint64(len(pid))))
	b.Write(pid)
	b.Write(varint.ToUvarint(uint64(len(v.ContextID))))
	b.Write(v.ContextID)
	b.Write(varint.ToUvarint(uint64(len(v.MetadataBytes))))
	b.Write(v.MetadataBytes)
	return b.Bytes(), nil
```